edgelet/edgelet-docker/src/image_prune_data.rs

```diff
@@ -0,0 +1,635 @@
+
+        // write previously removed entries back to file
+        if let Err(e) = write_images_with_timestamp(
+            &carry_over,
+            guard.tmp_filepath.clone(),
+            guard.image_use_filepath.clone(),
+        ) {
+            log::warn!("Failed to update image auto pruning persistence file. File will be updated on next scheduled run. {}", e);
+        };
+
+        /* ============================== */
+
+        drop(guard);
+
+        // these are the images we need to prune; file has already been updated
+        Ok(images_to_delete)
+    }
+}
+
+/* ===================================== HELPER METHODS ==================================== */
+
+fn get_images_with_timestamp(
+    image_use_filepath: String,
+) -> Result<HashMap<String, Duration>, Error> {
+    if !std::path::Path::new(&image_use_filepath).exists() {
+        log::info!(
+            "Image garbage collection data file not found; creating file at: {}",
+            image_use_filepath.as_str()
+        );
+        let _file = fs::File::create(image_use_filepath.clone()).map_err(Error::CreateFile)?;
+    }
+
+    let res = fs::read_to_string(image_use_filepath);
+    if let Err(e) = res {
+        let msg = format!("Could not read image persistence data: {}", e);
+        log::error!("{msg}");
+        return Err(Error::FileOperation(msg));
+    }
+
+    let contents = res.expect("Reading image persistence data failed");
+
+    let mut image_map: HashMap<String, Duration> = HashMap::new(); // all image pruning data
+
+    // TL;DR: this dumps pruning data into the image_map, where
+    // Key: Image hash, Value: Timestamp when image was last used (in epoch)
+    let contents = contents
+        .lines()
+        .map(|line| line.split(' ').collect::<Vec<&str>>())
+        .map(|vec| (vec[0].to_string(), vec[1]));
+    for (k, v) in contents {
+        image_map.insert(
+            k,
+            Duration::from_secs(v.parse::<u64>().map_err(Error::ParseIntError)?),
+        );
+    }
+
+    Ok(image_map)
+}
+
+fn write_images_with_timestamp(
+    state_to_persist: &HashMap<String, Duration>,
+    temp_file: String,
+    image_use_filepath: String,
+) -> Result<(), Error> {
+    // write to a temp file and then rename/overwrite to image persistence file (to prevent file write failures or corruption)
+    let mut file = std::fs::File::create(temp_file.clone())
+        .expect("Could not create temporary persistence file");
+
+    for (key, value) in state_to_persist {
+        let image_details = format!("{} {}\n", key, value.as_secs());
+        let res = write!(file, "{}", image_details);
+        if res.is_err() {
+            let msg = format!(
+                "Could not write image:{} with timestamp:{} to store",
+                key,
+                value.as_secs()
+            );
+            return Err(Error::FileOperation(msg));
+        }
+    }
+
+    match fs::rename(temp_file, image_use_filepath) {
```

Comment:
I don't think you need another file to store temp data. Just generate the temp data into memory.

Reply:
It's not to _store_ temp data, per se. It's to avoid file write failing.

This temp file has the data we want to retain and the old state file ends up getting clobbered.

Reply:
You can still generate the new state in memory, then overwrite the old state file when done though. I still don't see a reason to do it in a temporary file.

Reply:
The new state _is_ generated in memory.

The only disconnect here I think is why clobbering has been chosen vis-a-vis overwriting. The only extra effort is the rename/clobber. 

It was done this way to avoid the most extreme of edge cases where a file write fails and causes us to lose state. Losing state might mean these images potentially never being cleaned up.

My personal perspective is that this is six of one, half a dozen of the other.